apm-agent-plugins/apm-es-restclient-plugin/src/main/java/co/elastic/apm/es/restclient/ElasticsearchRestClientInstrumentation.java

```diff
@@ -0,0 +1,121 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+package co.elastic.apm.es.restclient;
+
+import co.elastic.apm.bci.ElasticApmInstrumentation;
+import co.elastic.apm.impl.transaction.Span;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.method.MethodDescription;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+import org.apache.http.Header;
+import org.elasticsearch.client.Request;
+import org.elasticsearch.client.Response;
+import org.elasticsearch.client.ResponseException;
+
+import javax.annotation.Nullable;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Map;
+
+import static net.bytebuddy.matcher.ElementMatchers.named;
+import static net.bytebuddy.matcher.ElementMatchers.takesArgument;
+import static net.bytebuddy.matcher.ElementMatchers.takesArguments;
+
+/**
+ * Instrumentation for Elasticsearch RestClient, currently supporting only synchronized queries.
+ * All sync operations go through org.elasticsearch.client.RestClient#performRequest(org.elasticsearch.client.Request)
+ */
+public class ElasticsearchRestClientInstrumentation extends ElasticApmInstrumentation {
+    private static final String ES_REST_CLIENT_INSTRUMENTATION_GROUP = "elasticsearch-restclient";
+    private static final String SPAN_TYPE = "es-restclient";
+
+    private static final String REST_CLIENT_CLASS_NAME = "org.elasticsearch.client.RestClient";
+    private static final String REQUEST_CLASS_NAME = "org.elasticsearch.client.Request";
+    private static final String PERFORM_REQUEST_METHOD_NAME = "performRequest";
+
+    static final String ELASTICSEARCH_NODE_KEY = "Elasticsearch-node";
+    static final String QUERY_STATUS_CODE_KEY = "Query-status-code";
+    static final String ERROR_REASON_KEY = "Error-reason";
+
+    @Advice.OnMethodEnter
+    private static void onBeforeExecute(@Advice.Argument(0) Request request,
+                                        @Advice.Local("span") Span span,
+                                        @Advice.Local("esre") ResponseException esre) {
+        if (tracer == null || tracer.getActive() == null) {
+            return;
+        }
+        span = tracer.getActive().createSpan()
+            .withType(SPAN_TYPE)
+            .appendToName(request.getEndpoint()).appendToName(" ").appendToName(request.getMethod())
+            .activate();
+
+
+        // Add request parameters to query info
+        for (Map.Entry<String, String> param: request.getParameters().entrySet()) {
+            span.addTag(param.getKey(), param.getValue());
```

Comment:
That could be problematic, as the param keys could have a high cardinality, leading to a mapping explosion as all tags are indexed. See https://www.elastic.co/blog/found-crash-elasticsearch, https://www.elastic.co/guide/en/elasticsearch/reference/6.4/mapping.html

Reply:
Tags are mainly meant for user-defined metadata. The information added by agents are usually added to the JSON schema, which makes it easier to align on what kind of information should be collected across agents.

Reply:
Sure, as I said in the top comment, this is only to show what we have available. Only question is whether we want to make it configurable so that a user can opt to get those collected